sympy/series/formal.py

```diff
@@ -0,0 +1,1119 @@
+            rep2 = x
+            rep2b = -x0
+        temp = f.subs(x, rep)
+        result = _compute_fps(temp, x, 0, S.One, hyper, order, rational, full)
+        if result is None:
+            return None
+        return (result[0], result[1].subs(x, rep2 + rep2b),
+                result[2].subs(x, rep2 + rep2b),
+                result[3].subs(x, rep2 + rep2b))
+
+    if f.is_polynomial(x):
+        return None
+
+    #  extract x**n from f(x)
+    mul = S.One
+    if extract and f.free_symbols.difference(set([x])):
+        m = Wild('m')
+        n = Wild('n', exclude=[m])
+        f = f.factor().powsimp()
+        s = f.match(x**n*m)
+        if s[n]:
+            for t in Add.make_args(s[n]):
+                if t.has(Symbol):
+                    mul *= (x)**t
+        if mul is not S.One:
+            f = (f / mul)
+
+    f = f.expand()
+
+    #  Break instances of Add
+    #  this allows application of different
+    #  algorithms on different terms increasing the
+    #  range of admissible functions.
+    if isinstance(f, Add):
+        result = False
+        ak = sequence(S.Zero, (0, oo))
+        ind, xk = S.Zero, None
+        for t in Add.make_args(f):
+            res = _compute_fps(t, x, 0, S.One, hyper, order, rational, full,
+                               False)
+            if res:
+                if not result:
+                    result = True
+                    xk = res[1]
+                if res[0].start > ak.start:
+                    seq = ak
+                    s, f = ak.start, res[0].start
+                else:
+                    seq = res[0]
+                    s, f = res[0].start, ak.start
+                save = Add(*[z[0]*z[1] for z in zip(seq[0:(f - s)], xk[s:f])])
+                ak += res[0]
+                ind += res[2] + save
+            else:
+                ind += t
+        if result:
+            return ak, xk, ind, mul
+        return None
+
+    result = None
+
+    # from here on it's x0=0 and dir=1 handling
+    k = Dummy('k')
+    if rational:
+        result = rational_algorithm(f, x, k, order, full)
+
+    if result is None and hyper:
+        result = hyper_algorithm(f, x, k, order)
+
+    if result is None:
+        return None
+
+    ak = sequence(result[0], (k, result[2], oo))
+    xk = sequence(x**k, (k, 0, oo))
+    ind = result[1]
+
+    return ak, xk, ind, mul
+
+
+def compute_fps(f, x, x0=0, dir=1, hyper=True, order=4, rational=True,
+                full=False):
+    """Computes the formula for Formal Power Series of a function.
+
+    Returns (sequence of coefficients, sequence of x, independent terms,
+    common terms).
+
+    Tries to compute the formula by applying the following techniques
+    (in order)
+
+    * rational_algorithm
+    * Hypergeomitric algorithm
+
+    Parameters
+    ==========
+
+    * x : Symbol, optional
+        If x is None and ``f`` is univariate, the univariate symbols will be
+        supplied, otherwise an error will be raised.
```

Comment:
`x` isn't optional in the current implementation. You could make it optional by making it a kwarg that defaults to `None`, which might make sense. Up to you.